src/uint/div.rs

```diff
@@ -1,548 +1,169 @@
 //! [`Uint`] division operations.
 
-use super::div_limb::{
-    Reciprocal, div_rem_limb_with_reciprocal, div2by1, div3by2, rem_limb_with_reciprocal,
-    rem_limb_with_reciprocal_wide,
-};
-use crate::{
-    CheckedDiv, ConstChoice, DivRemLimb, DivVartime, Limb, NonZero, RemLimb, Uint, Wrapping,
-    const_choice::u32_min,
-};
+use super::div_limb::Reciprocal;
+use crate::{CheckedDiv, DivRemLimb, DivVartime, Limb, NonZero, RemLimb, Uint, UintRef, Wrapping};
 use core::ops::{Div, DivAssign, Rem, RemAssign};
 
 use subtle::CtOption;
 
 impl<const LIMBS: usize> Uint<LIMBS> {
-    /// Computes `self / rhs` using a pre-made reciprocal,
-    /// returns the quotient (q) and remainder (r).
-    #[inline(always)]
+    /// Computes `self / rhs` using a pre-made reciprocal, returning the quotient
+    /// and remainder.
     pub const fn div_rem_limb_with_reciprocal(&self, reciprocal: &Reciprocal) -> (Self, Limb) {
-        div_rem_limb_with_reciprocal(self, reciprocal)
+        let mut quo = *self;
+        let rem = quo
+            .as_mut_uint_ref()
+            .div_rem_limb_with_reciprocal(reciprocal);
+        (quo, rem)
     }
 
-    /// Computes `self / rhs`, returns the quotient (q) and remainder (r).
-    #[inline(always)]
+    /// Computes `self / rhs`, returning the quotient and remainder.
     pub const fn div_rem_limb(&self, rhs: NonZero<Limb>) -> (Self, Limb) {
-        div_rem_limb_with_reciprocal(self, &Reciprocal::new(rhs))
+        let mut quo = *self;
+        let rem = quo.as_mut_uint_ref().div_rem_limb(rhs);
+        (quo, rem)
     }
 
     /// Computes `self % rhs` using a pre-made reciprocal.
-    #[inline(always)]
     pub const fn rem_limb_with_reciprocal(&self, reciprocal: &Reciprocal) -> Limb {
-        rem_limb_with_reciprocal(self, reciprocal)
+        self.as_uint_ref().rem_limb_with_reciprocal(reciprocal)
     }
 
-    /// Computes `self % rhs`.
-    #[inline(always)]
+    /// Computes `self % rhs` for a `Limb`-sized divisor.
     pub const fn rem_limb(&self, rhs: NonZero<Limb>) -> Limb {
-        rem_limb_with_reciprocal(self, &Reciprocal::new(rhs))
+        self.as_uint_ref().rem_limb(rhs)
     }
 
-    /// Computes `self` / `rhs`, returns the quotient (q) and the remainder (r)
+    /// Computes `self` / `rhs`, returning the quotient and the remainder.
     ///
     /// This function is constant-time with respect to both `self` and `rhs`.
     pub const fn div_rem<const RHS_LIMBS: usize>(
```

Comment:
This could simply delegate to the `div_rem` method of `UintRef`, but then performance suffers. There's little to no difference when the size-aware bitshift methods of `Uint` are used instead.

Reply:
Yes, I'm not sure why I couldn't get matching performance but I've seen this a few times now with the bit shifting methods. It seems to be a step too far for `Uint::<LIMBS>::as_uint_ref().nlimbs()` to be interpreted as a constant, whereas `LIMBS` is not a problem.

Reply:
Maybe try `#[inline(always)]`?

Reply:
It looks like the performance matches if `bounded_wrapping_shl_assign` is also inlined. I'll do a bit of testing to see if there's any negative impact from that.

Reply:
This is updated now. I added `rem_wide` support, which is actually more flexible than the old version as it does not need the low limbs to be the same width as the high limbs. This would allow `mul_mod(_vartime)` to support multiplication by a different number of limbs, although I haven't enabled that yet.